apps/designer/app/designer/shared/design-tokens-manager/design-tokens-manager.tsx

```diff
@@ -0,0 +1,178 @@
+    if (isOpen === false && errors.hasErrors) {
+      setErrors(initialErrors);
+    }
+  }, [isOpen]);
+
+  const handleChangeOrSubmit = (event: FormEvent<HTMLFormElement>) => {
+    event.preventDefault();
+    const formData = new FormData(event.currentTarget);
+    const data = Object.fromEntries(formData);
+    const errors = validate(tokens, data);
+
+    if (errors.hasErrors && event.type === "submit") {
+      setErrors(errors);
+    }
+
+    if (errors.hasErrors === false && event.type === "submit") {
+      onChangeComplete({ ...data, group, type } as DesignToken);
+      setIsOpen(false);
+    }
+  };
+
+  return (
+    <Popover modal open={isOpen} onOpenChange={setIsOpen}>
+      <PopoverTrigger
+        asChild
+        aria-label={isNew ? "Create Token" : "Edit Token"}
+      >
+        <Button
+          ghost
+          onClick={(event) => {
+            event.preventDefault();
+            setIsOpen(true);
+          }}
+        >
+          <PlusIcon />
+        </Button>
+      </PopoverTrigger>
+      <PopoverPortal>
+        <PopoverContent align="end" css={{ zIndex: "$zIndices$1" }}>
+          <form onChange={handleChangeOrSubmit} onSubmit={handleChangeOrSubmit}>
+            <Flex direction="column" gap="2" css={{ padding: "$spacing$7" }}>
+              <Label htmlFor="name">Name</Label>
+              <InputErrorsTooltip
+                errors={errors.name}
+                css={{ zIndex: "$zIndices$2" }}
+              >
+                <TextField id="name" name="name" />
+              </InputErrorsTooltip>
+              <Label htmlFor="value">Value</Label>
+              <InputErrorsTooltip
+                errors={errors.value}
+                css={{ zIndex: "$zIndices$2" }}
+              >
+                <TextField id="value" name="value" />
+              </InputErrorsTooltip>
+              <Label htmlFor="description">Description</Label>
+              <TextArea id="description" name="description" />
+              {isNew && (
+                <Button type="submit" variant="blue">
+                  Create
+                </Button>
+              )}
+            </Flex>
+          </form>
+          <PopoverHeader title={isNew ? "New Token" : name} />
+        </PopoverContent>
+      </PopoverPortal>
+    </Popover>
+  );
+};
+
+const TokenItem = ({ token }: { token: DesignToken }) => {
+  return <div>{token.value}</div>;
+};
+
+export const DesignTokensManager = ({ publish }: { publish: Publish }) => {
+  const [tokens, setTokens] = useDesignTokens();
+
+  return (
+    <>
+      {groups.map(({ group, type }) => {
+        return (
+          <CollapsibleSection
+            label={group}
+            key={group}
+            rightSlot={
+              <TokenEditor
+                group={group}
+                type={type}
+                onChangeComplete={(token) => {
+                  const nextTokens = [...tokens, token];
+                  publish({ type: "updateTokens", payload: nextTokens });
```

Comment:
I'd consider passing `token` instead of `nextTokens` here, which would allow to make patches more granular in `useUpdateTokens` (as I mentioned there)

Reply:
agree